### PR TITLE
Binary search tree

### PR DESCRIPTION
**Pull request by AdrianJunge**
Add abstract search tree and binary search tree

### CODE REVIEW

File: datastructures/search_trees/binary_search_tree.py [hunk 1]
```diff
@@ -0,0 +1,109 @@
+from datastructures.search_trees.search_tree import SearchTree, Node, T, V
+
+
+class BinarySearchTree(SearchTree):
+    def search(self, search_key: T, current_node: Node = None) -> tuple[Node]:
+        if current_node is None:
+            current_node = self.root
```
Comment (lupeterm) [inline datastructures/search_trees/binary_search_tree.py]:
```suggestion
        current_node = current_node or self.root
```

File: datastructures/search_trees/binary_search_tree.py [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+            if comparison_current_vs_search:
+                current_node = current_node.right
+            else:
+                current_node = current_node.left
+        return current_node
+
+
+    def insert(self, key: T, value: V, current_node: Node = None) -> None:
+        if not current_node:
+            current_node = self.root
```
Comment (lupeterm) [inline datastructures/search_trees/binary_search_tree.py]:
```suggestion
        current_node = current_node or self.root
```

File: datastructures/search_trees/binary_search_tree.py [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+                else:
+                    current_node.right = new_node
+                    new_node.parent = current_node
+            else:
+                if current_node.left:
+                    self.insert(key, value, current_node.left)
+                else:
+                    current_node.left = new_node
+                    new_node.parent = current_node
+        return
```
Comment (lupeterm) [inline datastructures/search_trees/binary_search_tree.py]:
```suggestion
        if not self.root:
            self.root = new_node
            return
        comparison = self.order_relation(current_node.key, key)
        if comparison:
            if current_node.right:
                self.insert(key, value, current_node.right)
            else:
                current_node.right = new_node
                new_node.parent = current_node
        else:
            if current_node.left:
                self.insert(key, value, current_node.left)
            else:
                current_node.left = new_node
                new_node.parent = current_node
        return
```

Comment (lupeterm) [reply datastructures/search_trees/binary_search_tree.py]: i.e. just reduce indentation

File: datastructures/search_trees/binary_search_tree.py [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+                search_from_node = found_node.right if found_node.right else found_node
+                min_node, _ = self._get_min_by_node(search_from_node)
+                if min_node.parent != found_node:
+                    self._shift_nodes(min_node, min_node.right) # there can only be right nodes
+                    min_node.right = found_node.right
+                    min_node.right.parent = min_node
+                self._shift_nodes(found_node, min_node)
+                min_node.left = found_node.left
+                min_node.left.parent = min_node
+        return found_value
```
Comment (lupeterm) [inline datastructures/search_trees/binary_search_tree.py]: same here as above, check negative case first and return, then reduce indentation for other case

File: datastructures/search_trees/binary_search_tree.py [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+
+
+    def visualize(self) -> str:
+        lines, *_ = self._display_aux()
```
Comment (lupeterm) [inline datastructures/search_trees/binary_search_tree.py:62]: what in gods name kind of syntax is  `VAR, *_`

File: datastructures/search_trees/binary_search_tree.py [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+        for line in lines:
+            print(line)
+
+
+    def _display_aux(self, current_node = None):
+        if not current_node:
+            current_node = self.root
```
Comment (lupeterm) [inline datastructures/search_trees/binary_search_tree.py]:
```suggestion
        current_node = current_node or self.root
```

File: datastructures/search_trees/search_tree.py [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+        self.order_relation = order_relation
+        self.tree_size = 0
+        self.root = None            
+
+
+    def _get_min_by_node(self, current_node) -> tuple[Node, V]:
+        if current_node is not None:
+            while current_node.left:
+                current_node = current_node.left
+        return current_node, current_node.value
```
Comment (lupeterm) [inline datastructures/search_trees/search_tree.py:30]: könnte das hier nicht potentiell nen NonetypeError werfen?

**Review by lupeterm — DISMISSED** [review]

File: datastructures/search_trees/binary_search_tree.py (hunk 1 above)
Comment (AdrianJunge) [reply datastructures/search_trees/binary_search_tree.py]: That looks good, didn't know this is a thing

File: datastructures/search_trees/binary_search_tree.py (hunk 5 above)
Comment (AdrianJunge) [reply datastructures/search_trees/binary_search_tree.py:62]: Everything being returned behind VAR is just discarded.
Btw this code visualize is from Stackoverflow :D

File: datastructures/search_trees/search_tree.py (hunk 7 above)
Comment (AdrianJunge) [reply datastructures/search_trees/search_tree.py:30]: Yes you're right 👍

File: datastructures/search_trees/binary_search_tree.py (hunk 5 above)
Comment (lupeterm) [reply datastructures/search_trees/binary_search_tree.py:62]: Ja ach was als ob du selber so eine atrocity fabrizieren würdest xd

**Review by lupeterm — APPROVED** [review]

**Pull request merged**